tests/test_paragraph_hash.py

```
@@ -0,0 +1,418 @@
+
+    def test_parse_large_index(self):
+        ref = ParagraphRef.parse("P999#ffff")
+        assert ref.index == 999
+        assert ref.hash == "ffff"
+
+    def test_reject_invalid_format(self):
+        with pytest.raises(ValueError, match="Invalid paragraph reference"):
+            ParagraphRef.parse("paragraph3")
+
+    def test_reject_missing_hash(self):
+        with pytest.raises(ValueError):
```

Ruff (0.15.7)

[warning] `pytest.raises(ValueError)` is too broad, set the `match` parameter or use a more specific exception

(PT011)